### PR TITLE
[MM-19720] Expose endpoint for config patch

### PR DESCRIPTION
**Pull request by RajatVaryani**
#### Ticket Link
Fixes https://github.com/mattermost/mattermost-server/issues/12958

### CODE REVIEW

**Comment by jwilander** [comment]
@RajatVaryani feel free to mention me when it's ready for review

**Comment by jasonblais** [comment]
Hey @RajatVaryani, let us know if you have any questions. We'd be happy to help! :)

**Comment by RajatVaryani** [comment]
@jasonblais I will be closing this issue by weekend. There are no issues with the ticket.

**Comment by RajatVaryani** [comment]
While working I found a redundant check in the code https://github.com/mattermost/mattermost-server/blob/9b6a0674f71cdbfe510792d716b6b630fbb9f814/api4/config.go#L82. As we have restrict already true in the struct https://github.com/mattermost/mattermost-server/blob/9b6a0674f71cdbfe510792d716b6b630fbb9f814/model/config.go#L2239. What you think should I remove it?

**Comment by RajatVaryani** [comment]
@jwilander This is ready for review. I have the following questions/suggestions:

1.  Do we have any strategy for arranging functions in a file?
2. Should we  make use of `Patch` HTTP verb instead of `PUT`?
3. Should we also move the special logic associated with message export config to a layer blow I.e in app layer as it is the ideal place for it? If we do so testing will also become easier.

**Comment by jwilander** [comment]
Awesome @RajatVaryani we'll get it reviewed. To answer your questions:

1. Not really, we tend to add new functions around where they make sense but we have not strict ordering that we adhere to
2. No, we should use `PUT` to conform with all our other patch endpoints that use `PUT`
3. Yes, I would suggest moving any extra or special logic that is not directly involved with reading/interpreting the request and building a response into the app layer 👍

**Comment by cpoile** [comment]
Hi @RajatVaryani, will you be making any adjustments based on jwilander's feedback? (Because I'd like to hold off reviewing until you're ready.) Thanks!

**Comment by RajatVaryani** [comment]
Yes, Please. I will be making some changes. I will tag you once done. 

Thank you.

**Comment by RajatVaryani** [comment]
@cpoile Please review it. I am done with the changes.

File: api4/config_test.go [hunk 1]
```diff
@@ -342,3 +342,86 @@ func TestGetOldClientConfig(t *testing.T) {
...
+		*th.App.Config().ExperimentalSettings.RestrictSystemAdmin = false
+
+		config := model.Config{PasswordSettings: model.PasswordSettings{
+			Lowercase: model.NewBool(true),
+			Number:    model.NewBool(true),
+			Uppercase: model.NewBool(true),
+			Symbol:    model.NewBool(true),
+		}, LogSettings: model.LogSettings{
+			ConsoleLevel: model.NewString("INFO"),
+		}}
```
Comment (cpoile) [inline api4/config_test.go]: Just wondering if we could be explicit about:
1. Testing a before and after the patch
2. Patching a few different structs in the config
3. Maybe a couple that are nested a level or two
4. And maybe a couple of different types (string, int, slice)

Comment (RajatVaryani) [reply api4/config_test.go]: Fair point. I have made changes. Before patching the config I make sure what are the values in config then I call patch and once again at the end make sure the values have changed. I have added added some different structs. Please have a look.

**Comment by cpoile** [comment]
@DHaussermann Not sure if we should test this, or who should take a look at this -- would require them to use the endpoint?

**Comment by RajatVaryani** [comment]
> @RajatVaryani Nice work, thanks again for the contribution!

@cpoile Thank you for the quick turnover time and the opportunity. :)

**Comment by DHaussermann** [comment]
@cpoile I might be misunderstanding something here but this seems simple enough to test if you can provide a valid payload for CURL or json, I could then use a curl command or Postman to ensure I get  the expected responses from the end point and the config is applied.

**Comment by cpoile** [comment]
@DHaussermann Awesome, sounds great, thanks!

**Comment by RajatVaryani** [comment]
@cpoile Should I make changes to the API documentation?

**Comment by DHaussermann** [comment]
@cpoile can you provider a sample payload that I can use to test this?

**Comment by RajatVaryani** [comment]
> @cpoile can you provider a sample payload that I can use to test this?

@DHaussermann I can provide you that. Please feel free to get in touch if you have any other questions. I am @rajatv on Mattermost community.

**Comment by cpoile** [comment]
@RajatVaryani That would be great (add docs), if you're comfortable making a doc PR? Let me know if you need help.

**Comment by jasonblais** [comment]
@DHaussermann Just touching base to see if you have what you need to carry out QA review for this PR?

**Comment by DHaussermann** [comment]
I tried testing out the new end point on this ☝️ spin-up server but, I don't see this working.
@RajatVaryani or @cpoile can you please provide some guidance on this? 

I tried testing this will CURL and Postman. The request does not time out but seems to send a blank reply. Almost as though the end point is not valid. 

Config_EndPoint

I'm not sure what I'm doing wrong. I am sending the request to `'https://mattermost-server-pr-12997.test.mattermost.cloud/config/patch'` as a `PUT`. I'm using a sample payload provided by @RajatVaryani but, given that I don't get any meaningful reply I'm guessing the problem is not related to the request body.

**Comment by RajatVaryani** [comment]
@DHaussermann @cpoile  The url should be `https://mattermost-server-pr-12997.test.mattermost.cloud/api/v4/config/patch`. I tried but now I am getting `{"id":"api.context.404.app_error","message":"Sorry, we could not find the page.","detailed_error":"There doesn't appear to be an api call for the url='/api/v4/config/patch'.  Typo? are you missing a team_id or user_id as part of the url?","status_code":404}` as response. The server returns this error only when the given api endpoint does not exists. But I changed the baseUrl to localhost and it is working fine. Is it possible that the test server spinned up was on an older version or without all my commits?

**Comment by RajatVaryani** [comment]
@DHaussermann It is working now with ```curl --location --request PUT 'https://mattermost-server-pr-12997.test.mattermost.cloud/api/v4/config/patch' \
--header 'Authorization: Bearer rgb9tm4z338j3buw3dsfqd95ke' \
--header 'Content-Type: application/json' \
--header 'X-Requested-With: XMLHttpRequest' \
--data-raw '{
    "PluginSettings": {
        "Enable": true,
        "EnableUploads": true
    }}'```

Another problem is the `PUT {{baseURL}}/config` is not working. How can I get a test server for master branch so that I can confirm if it is an issue with my PR or something else?

**Comment by DHaussermann** [comment]
Thanks @RajatVaryani! I did not realize `api/v4`was missing.

I did successfully test that updates to the the config file are working using CURL and Postman. I did not have any issue using Postman with LDAP enabled and the `GuestFilter` empty.  Not sure if there was an extra step needed to repro this?

I tested that one and multiple values can updated in the confg. I also spot-checked various values and confirmed that you can send a requests which updates specific values from different sections of the config successfully. 

Interestingly, the `EnableUploads` value you provided in the example was the only one where I saw a strange behaviour. It cannot be set to `false`. This value remains unchanged when I include it in the requests. It seems strange that updates would not be made to the config for a specific value when others are working. Could there be something else happening on the server preventing this from updating? I did not see anything relevant in the logs.

**Comment by RajatVaryani** [comment]
@DHaussermann Please have a look at https://api.mattermost.com/#tag/system%2Fpaths%2F~1config%2Fput. We cannot toggle `EnableUploads` in config update. The same behaviour is maintained in patch as well. I confirmed this behaviour [here](https://github.com/mattermost/mattermost-server/issues/12958#issuecomment-549113592)

**Comment by cpoile** [comment]
> Another problem is the `PUT {{baseURL}}/config` is not working from the postman. That is because the default request payload is not having `"GuestFilter": ""` in `LdapSettings` block. We should also update that.

Hi @RajatVaryani, can you explain a bit more what you mean by this?

**Comment by DHaussermann** [comment]
@RajatVaryani the testing done on the end point found no issues. I believe we're okay to merge this. It's just that I'm not sure what your follow up concern was. 

Can you please advise if this is something that can be addressed in a follow up issue so we can merge this PR?

**Comment by RajatVaryani** [comment]
@DHaussermann We can merge this PR. Missed replying to your query.

**Pull request merged**